File: apps/backend/app/models/scene.py
"""
Scene 模型 — 章节分场（三层调度核心）

每章拆分为 4-8 场（scene），每场独立规划：
- POV、时间、地点、在场角色、目标、冲突、转折、钩子、字数预算、感官焦点、节奏

这是把“AI 味平”根治的工程性关键：模型不再“展开式”写整章，而是按分场逐场写、正文 stitch、自检。

与 OutlineNode 的关系：
- 一章通常对应 1 个 OutlineNode，但可拆多个 Scene（order 1..N）
- Scene 可独立于 OutlineNode（作者手动调整分场时）
"""

from sqlalchemy import Column, String, Text, DateTime, ForeignKey, Integer, JSON
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import uuid
from app.database import Base


class Scene(Base):
    __tablename__ = "scenes"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)  # 主键
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id"), nullable=False)  # FK → projects.id
    chapter_id = Column(UUID(as_uuid=True), ForeignKey("chapters.id"), nullable=True)  # FK → chapters.id；成稿后绑定章节
    outline_node_id = Column(UUID(as_uuid=True), ForeignKey("outline_nodes.id"), nullable=True)  # FK → outline_nodes.id，对应大纲节点

    order = Column(Integer, nullable=False)           # 场次 1..N
    title = Column(String(200))                       # 场标题（可选）

    # 时间线
    time = Column(String(100))                        # “第3日·夜” 或 “同日·上午”
    story_day = Column(String(50))                    # 故事内日期表述
    location_id = Column(UUID(as_uuid=True), ForeignKey("locations.id"), nullable=True)  # FK → locations.id；由 main.py 兼容迁移补齐
    location_name = Column(String(200))               # 冗余保留，方便旧数据与 free-text 兜底

    # 视点与角色
    pov_character_id = Column(UUID(as_uuid=True), ForeignKey("characters.id"), nullable=True)  # FK → characters.id，本场 POV
    characters_on_stage = Column(JSON, default=list)  # [char_id, ...] 本场在场角色

    # 戏剧元素
    goal = Column(Text)                               # 本场目标（角色想要什么）
    conflict = Column(Text)                           # 冲突/障碍
    turn = Column(Text)                               # 转折（本场发生的关键变化）
    hook = Column(Text)                               # 场末钩子（留给下一场或章末）
    hook_strength = Column(Integer, default=3)        # 1-5，钩子强度

    # 写作约束
    word_budget = Column(Integer, default=400)        # 预算字数
    actual_word_count = Column(Integer, default=0)
    pacing = Column(String(20), default="mid")        # fast / mid / slow
    sensory_focus = Column(String(30), default="mixed")  # sight / sound / smell / taste / touch / mixed

    # 状态
    status = Column(String(20), default="planned")    # planned / written / reviewed
    content = Column(Text)                            # 本场正文（写完后存）

    # ── 投料约束区（由 chapter_ingredients 服务在分场规划时写入）────
    # 本场需推进的故事线：[{storyline_id, name, line_type, from_state, to_state, must_advance}]
    storyline_moves = Column(JSON, default=None)
    # 势力/地盘着色：{faction_id, name, faction_type, alignment, atmosphere, npc_default_attitude}
    faction_color = Column(JSON, default=None)
    # 技能/法宝聚光灯：[{asset_type, asset_id, name, description, key_effect}]
    asset_spotlight = Column(JSON, default=None)
    # 伏笔操作指令：[{foreshadow_id, title, op, suggested_method, is_overdue}]
    foreshadow_ops = Column(JSON, default=None)
    # 债务标记：[{debt_type, description, severity, overdue_chapters}]
    debt_flags = Column(JSON, default=None)
    # 结构预警（规划时生成）：[{code, level, msg}]
    structural_warnings = Column(JSON, default=None)
    # 写后核验结果（stitch 后异步回填）：{storyline_ok, foreshadow_ok, debt_cleared, score, notes}
    checklist_result = Column(JSON, default=None)

    extra = Column(JSON, default=dict)  # JSON 扩展字段
    created_at = Column(DateTime(timezone=True), server_default=func.now())  # 创建时间（UTC）
    updated_at = Column(DateTime(timezone=True), onupdate=func.now(), server_default=func.now())  # 更新时间（UTC）

    project = relationship("Project", back_populates="scenes")
    chapter = relationship("Chapter", back_populates="scenes")
    outline_node = relationship("OutlineNode", back_populates="scenes")
    pov_character = relationship("Character", foreign_keys=[pov_character_id])
    location = relationship("Location", back_populates="scenes", foreign_keys=[location_id])

File: apps/client/src/types/scene.ts
// ── Scene（三层调度：章纲 → 分场 → 正文） ──────────────────

/** 节奏标记 */
export type ScenePacing = 'fast' | 'mid' | 'slow'

/** 分场状态 */
export type SceneStatus = 'planned' | 'written' | 'reviewed'

// ── 场景投料约束类型 ──────────────────────────────────────
export interface SceneStorylineMove {
  storyline_id: string
  name: string
  line_type: string
  must_advance: boolean
  gap_chapters: number
  suggested_beat?: string
}

export interface SceneDebtFlag {
  debt_type: 'payoff' | 'storyline_gap' | 'promise_due' | 'emotion_debt' | string
  description: string
  severity: 'critical' | 'warning' | 'info'
  overdue_chapters: number
  related_id?: string
}

export interface SceneForeshadowOp {
  foreshadow_id: string
  title: string
  op: 'lay' | 'hint' | 'resolve'
  priority: number
  suggested_method: string
  is_overdue: boolean
}

export interface SceneFactionColor {
  faction_id: string
  name: string
  faction_type: string
  alignment: string
  atmosphere: string
  npc_default_attitude: string
}

export interface SceneAssetCard {
  asset_type: 'skill' | 'item'
  asset_id: string
  name: string
  description: string
  key_effect: string
  cost_or_rarity?: string
}

export interface SceneStructuralWarning {
  code: string
  level: 'warn' | 'info'
  msg: string
}

export interface SceneChecklistResult {
  storyline_ok: boolean
  foreshadow_ok: boolean
  debt_cleared: boolean
  score: number
  notes: string
}

/**
 * 分场（Scene）——章节的子结构单元。
 * 对应后端 SceneRead schema；由 Bootstrap Step 13 或手动生成。
 * outline_node_id 指向所属 chapter_plan 大纲节点。
 */
export interface Scene {
  id: string
  project_id: string
  outline_node_id: string | null
  chapter_id: string | null
  order: number
  title: string | null
  time: string | null
  story_day: string | null
  /** 关联 Location 库记录的 UUID；存在时写章会注入感官基准约束块 */
  location_id: string | null
  location_name: string | null
  pov_character_id: string | null
  /** 在场人物 ID 列表 */
  characters_on_stage: string[]
  goal: string | null
  conflict: string | null
  turn: string | null
  hook: string | null
  hook_strength: number
  word_budget: number
  /** 实际已写字数（后端 scene_draft/stream 写完后更新） */
  actual_word_count: number
  pacing: ScenePacing
  sensory_focus: string
  status: SceneStatus
  content: string | null
  extra: Record<string, unknown>
  // ── 投料约束字段（分场规划时由 chapter_ingredients 服务写入）────
  storyline_moves: SceneStorylineMove[] | null
  debt_flags: SceneDebtFlag[] | null
  foreshadow_ops: SceneForeshadowOp[] | null
  faction_color: SceneFactionColor | null
  asset_spotlight: SceneAssetCard[] | null
  structural_warnings: SceneStructuralWarning[] | null
  /** 写后核验结果（stitch 后异步回填） */
  checklist_result: SceneChecklistResult | null
}
